hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/BlobListXmlParser.java

```diff
@@ -121,57 +149,136 @@ public void endElement(final String uri,
       currentBlobProperty = null;
     }
 
+    /*
+     * If the closing tag is BlobPrefix, there are no more properties to be set in
+     * currentBlobProperty and this is a directory (implicit or explicit)
+     * If implicit, it will be added only once/
+     * If explicit it will be added with Blob Tag as well.
+     */
+    if (AbfsHttpConstants.BLOB_PREFIX.equals(currentNode)) {
+      currentBlobProperty.setIsDirectory(true);
+      blobList.addBlobProperty(currentBlobProperty);
+      currentBlobProperty = null;
+    }
+
+    /*
+     * If the closing tag is Next Marker, it needs to be saved with the
+     * list of blobs currently fetched
+     */
     if (AbfsHttpConstants.NEXT_MARKER.equals(currentNode)) {
       blobList.setNextMarker(value);
     }
 
-    if (parentNode.equals(AbfsHttpConstants.BLOB_PREFIX)) {
-      if (currentNode.equals(AbfsHttpConstants.NAME)) {
-        currentBlobProperty.setBlobPrefix(value);
-      }
-    }
     /*
-     * For case:
-     * <Blob>
-     * <Name>value</name>
-     * ....</Blob>
+     * If the closing tag is Name, then it is either for a blob
+     * or for a blob prefix denoting a directory. We will save this
+     * in current BlobProperty for both
      */
-    if (parentNode.equals(AbfsHttpConstants.BLOB)) {
-      if (currentNode.equals(AbfsHttpConstants.NAME)) {
-        currentBlobProperty.setName(value);
-        currentBlobProperty.setPath(new Path(ROOT_PATH + value));
-        currentBlobProperty.setUrl(url + ROOT_PATH + value);
+    if (currentNode.equals(AbfsHttpConstants.NAME)) {
+      // In case of BlobPrefix Name will have a slash at the end
+      // Remove the "/" at the end of name
+      if (value.endsWith(FORWARD_SLASH)) {
+        value = value.substring(0, value.length() - 1);
```

Comment:
lets still have parentNode check. Somthing like
`parentNode.equals(BLOB) || parentNode.equals(BLOB_PREFIX)`

reason being, NAME can be there in other tags as well like Metadata.

Reply:
Makes sense
Taken